Expose additional app definition variables

wendigo opened the pull request:
I would like to expose additional appDefinition fields as environment variables

gkleiman commented:
LGTM

wendigo commented:
Waiting for comments then :)

gkleiman commented on docs/docs/task-environment-vars.md:
```diff
@@ -44,6 +44,10 @@ names through the `--env_vars_prefix` command line flag.
 - `MARATHON_APP_DOCKER_IMAGE` contains the value of the app definition 
   `container.docker.image` value. For example `mesosphere/marathon:latest`, `nginx` or `nginx:1.9.3`.
+- `MARATHON_APP_CPUS` contains the value of the app definition `cpus` value.
+- `MARATHON_APP_MEMORY` contains the value of the app definition `mem` value, expressed in megabytes.
```
Can we please follow the naming schema used in the API, i.e., 'MEM' and 'DISK'?

wendigo replied:
OK.

gkleiman commented on docs/docs/task-environment-vars.md:
```diff
@@ -44,6 +44,10 @@ names through the `--env_vars_prefix` command line flag.
+- `MARATHON_APP_CPUS` contains the value of the app definition `cpus` value.
+- `MARATHON_APP_MEMORY` contains the value of the app definition `mem` value, expressed in megabytes.
+- `MARATHON_APP_DISK_SPACE`  value of the app definition `disk` value, expressed in megabytes.
+- `MARATHON_APP_LABELS` contains list of labels of the corresponding app definition. For example: `label1:value1;label2:value2`
```
The following variables would be easier to use from shell scripts:

MARATHON_APP_LABELS=“LABEL1 LABEL2 LABEL3”
MARATHON_APP_LABEL_LABEL1=value1
MARATHON_APP_LABEL_LABEL2=value2

We would however have to make sure we convert invalid characters/spaces in the label names to underscores.

wendigo replied:
OK. Will do

wendigo replied:
I will follow discussion here: http://stackoverflow.com/questions/2821043/allowed-characters-in-linux-environment-variable-names to provide environment variable name safety

aquamatthias commented:
Hey @wendigo thanks for this PR.

1) Resources
Please change the names of the resources to MARATHON_APP_RESOURCE_*:
MARATHON_APP_RESOURCE_MEM
MARATHON_APP_RESOURCE_CPUS
MARATHON_APP_RESOURCE_DISK

2) Labels
There is a size limit for env vars: See here: http://man7.org/linux/man-pages/man2/execve.2.html
"Limits one size of arguments and environment"
I guess it makes sense to only include labels, that do not exceed a defined length (key and value).
We should make that length adjustable (default 512 chars?).

Thanks again! Matthias

wendigo commented:
Hi @aquamatthias:

1) Resources - I've changed variable names according to your comment
2) Implemented. Limit (512) is hardcoded for now. I think I can manage to make it configurable if you would like to.

P.S. I'm not a Scala dev so all comments are appreciated :)

wendigo commented on src/main/scala/mesosphere/mesos/TaskBuilder.scala:
```diff
@@ -1,6 +1,6 @@
 package mesosphere.mesos
 
-import com.google.protobuf.{ TextFormat, ByteString }
+import com.google.protobuf.{ ByteString, TextFormat }
```
IntelliJ :)

aquamatthias commented on docs/docs/task-environment-vars.md:
```diff
@@ -44,6 +44,11 @@ names through the `--env_vars_prefix` command line flag.
   was used to start this task. For example `2015-04-02T09:37:00.596Z`.
 - `MARATHON_APP_DOCKER_IMAGE` contains the value of the app definition 
   `container.docker.image` value. For example `mesosphere/marathon:latest`, `nginx` or `nginx:1.9.3`.
+- `MARATHON_RESOURCE_CPUS` contains the value of the app definition `cpus` value.
```
Please use MARATHON_APP_RESOURCE_CPUS

aquamatthias commented on docs/docs/task-environment-vars.md:
```diff
@@ -44,6 +44,11 @@ names through the `--env_vars_prefix` command line flag.
 - `MARATHON_APP_DOCKER_IMAGE` contains the value of the app definition 
   `container.docker.image` value. For example `mesosphere/marathon:latest`, `nginx` or `nginx:1.9.3`.
+- `MARATHON_RESOURCE_CPUS` contains the value of the app definition `cpus` value.
+- `MARATHON_RESOURCE_MEM` contains the value of the app definition `mem` value, expressed in megabytes.
```
Please use MARATHON_APP_RESOURCE_MEM

aquamatthias commented on docs/docs/task-environment-vars.md:
```diff
@@ -44,6 +44,11 @@ names through the `--env_vars_prefix` command line flag.
   `container.docker.image` value. For example `mesosphere/marathon:latest`, `nginx` or `nginx:1.9.3`.
+- `MARATHON_RESOURCE_CPUS` contains the value of the app definition `cpus` value.
+- `MARATHON_RESOURCE_MEM` contains the value of the app definition `mem` value, expressed in megabytes.
+- `MARATHON_RESOURCE_DISK`  value of the app definition `disk` value, expressed in megabytes.
```
Please use MARATHON_APP_RESOURCE_DISK

aquamatthias commented on src/main/scala/mesosphere/mesos/TaskBuilder.scala:
```diff
@@ -318,10 +321,38 @@ object TaskBuilder {
+  }
+
+  def escapeEnvironmentVariableName(name: String): String = {
+    val variableName = name.replaceAll("[^a-zA-Z0-9_]+", "_").toUpperCase
```
This would replace subsequent invalid chars into a single underscore: e.g. `1234$$$$` into `1234_`. If you omit the leading `+`, then the string would be replaced to `1234____`
I am not exactly sure which one is better...

aquamatthias commented on src/main/scala/mesosphere/mesos/TaskBuilder.scala:
```diff
@@ -318,10 +321,38 @@ object TaskBuilder {
+    val variableName = name.replaceAll("[^a-zA-Z0-9_]+", "_").toUpperCase
+
+    if (variableName.length > maxVariableLength) {
+      variableName.substring(0, maxVariableLength)
```
I rather would completely omit a variable if it is longer than the maximum.
Truncating could lead to strange behavior as well...

wendigo replied:
Ok :)

aquamatthias commented on src/main/scala/mesosphere/mesos/TaskBuilder.scala:
```diff
@@ -318,10 +321,38 @@ object TaskBuilder {
+
+    labels.foreach {
+      case (key, value) => {
+        variables = variables ++ Map(s"${labelEnvironmentKeyPrefix}${escapeEnvironmentVariableName(key)}" -> value)
```
What about the values. They could exceed size as well, or?

aquamatthias replied:
You can do something like, so you don't create a new map with every entry:
```scala
variables += key->value
```

aquamatthias commented on src/main/scala/mesosphere/mesos/TaskBuilder.scala:
```diff
@@ -318,10 +321,38 @@ object TaskBuilder {
+        "MARATHON_APP_DOCKER_IMAGE" -> app.container.flatMap(_.docker.map(_.image)),
+        "MARATHON_RESOURCE_CPUS" -> Some(app.cpus.toString),
+        "MARATHON_RESOURCE_MEM" -> Some(app.mem.toString),
+        "MARATHON_RESOURCE_DISK" -> Some(app.disk.toString)
```
See naming scheme

wendigo replied:
My mistake. Adjusted naming as you suggested.

aquamatthias commented on src/main/scala/mesosphere/mesos/TaskBuilder.scala:
```diff
@@ -318,10 +322,38 @@ object TaskBuilder {
+    )
+
+    labels.foreach {
+      case (key, value) => {
```
You can check the length of key and value here:
`case (key, value) if key.length<512 && value.length<512 => ...`

aquamatthias commented on docs/docs/task-environment-vars.md:
```diff
@@ -44,10 +44,18 @@ names through the `--env_vars_prefix` command line flag.
 specified via the `--env_vars_prefix` command line flag.
 
+_Note: Label name will be sanitized: all characters except a-z A-Z 0-9 _ will be replaced with _ and 
+label will be truncated to 512 chars (by default).
```
Not truncated, but filtered out.

aquamatthias commented:
@wendigo Thanks for the update. 
Please also check the length of the label value. 
If you could squash the commits, I can merge the functionality.

wendigo commented:
Commit squashed.

aquamatthias commented on src/main/scala/mesosphere/mesos/TaskBuilder.scala:
```diff
@@ -318,10 +323,42 @@ object TaskBuilder {
+    }
+  }
+
+  def labelsToEnvVars(labels: Map[String, String]): Map[String, String] = {
```
Sorry to be picky: collect already filters automatically. May I suggest the following code:
```scala
def labelsToEnvVars(labels: Map[String, String]): Map[String, String] = {
  def escape(name: String): String = name.replaceAll("[^a-zA-Z0-9_]+", "_").toUpperCase
  val validLabels = labels.collect {
    case (key, value) if key.length<maxEnvLen && value.length<maxEnvLen => escape(key) -> value
  }

  val names = Map("MARATHON_APP_LABELS" -> validLabels.keys.mkString(" "))
  val values = validLabels.map { case (key, value) => s"$labelEnvironmentKeyPrefix$key" -> value }
  names ++ values
}
```

wendigo replied:
Done :)

aquamatthias replied:
Great - will wait for the tests.

wendigo replied:
Passed :)

aquamatthias commented:
Thanks!

lloesche commented:
@wendigo I love this feature, thank you!
